world/traits.py

```diff
@@ -0,0 +1,701 @@
+# -*- coding: utf-8 -*-
+"""
+Traits Module
+
+`Trait` classes represent modifiable traits on objects or characters. They
+are instantiated by a `TraitFactory` object, which is typically set up
+as a property on the object or character's typeclass.
+
+Data persistence for `Trait` objects is handled by passing a dict from
+a typeclass instance's `db` attribute handler loaded with the appropriate
+keys to the factory which map to properties of the eventual `Trait` object.
+See _Trait Configuration_ below for more details.
+
+**Setup**
+    To use traits on an object, add a dict containing trait configuration
+    keys to the object's `db` property in the `at_object_creation` hook,
+    and define a property on the object's class that passes that Attribute
+    into the constructor and returns a `TraitFactory`.
+
+Example:
+    ```python
+    from world.traits import TraitFactory
+        ...
+    class Object(DefaultObject):
+        ...
+        def at_object_creation(self):
+            self.db.traits = {
+                # static trait example
+                'str': {'name': 'Strength',
+                        'type': 'static',
+                        'base': 5},
+                # counter trait example
+                'carry': {'name': 'Carry Weight',
+                          'type': 'counter',
+                          'base': 0,
+                          'min': 0,
+                          'max': None},
+                # gauge trait example
+                'hp': {'name': 'HP',
+                       'type': 'gauge',
+                       'base': 10}
+            }
+        ...
+        @property
+        def traits(self):
+            return TraitFactory(self.db.traits)
+    ```
+
+**Trait Configuration**
+    `Trait` objects can be configured as one of three basic types with
+    increasingly complex behavior.
+
+    * Static - A simple trait model with a base value and optional modifier.
+
+    * Counter - Trait with a base value and a modifiable current value that
+        can vary along a range defined by optional min and max values.
+
+    * Gauge - Modified counter type modeling a refillable "gauge".
+
+    All traits have a read-only `actual` property that will report the trait's
+    actual value. In addition to normal attribute notation, `Trait` objects
+    support the unary `+` operator as a shortcut to access this property.
+
```

Comment:
The use cases were driven both by OA rules and after consideration of common categories of trait behaviors I've seen in many games. There are examples of all three trait types in the base trait configuration in `__init__` for the `Archetype` class.

In working on this, I was informed by Ainneve's use cases, but was also trying to make the `Trait` class something general purpose and robust enough that it could be easily reused in others' games. That's also part of why I went so over the top with the documentation.